File: src/test/java/org/oewntk/model/LibTestModelLexGroups.java
/*
 * Copyright (c) 2021. Bernard Bou.
 */

package org.oewntk.model;

import java.io.PrintStream;
import java.io.StringWriter;
import java.util.List;
import java.util.Locale;

import static org.junit.Assert.assertEquals;
import static org.oewntk.model.SenseGroupings.sensesForLCLemma;
import static org.oewntk.model.SenseGroupings.sensesForLCLemmaAndPos;

public class LibTestModelLexGroups
{
	public static void testCIMultipleAll(final CoreModel model, final PrintStream ps)
	{
		LexGroupings.lemmasByICLemmaHavingMultipleCount(model) //
				.forEach((u, cs) -> ps.printf("%s {%s}%n", u, String.join(",", cs)));
	}

	public static void testCILemmas(final CoreModel model, final String word, final PrintStream ps)
	{
		var lemmas = LexGroupings.lemmasByICLemma(model).get(word);
		ps.printf("%s {%s}%n", word, String.join(",", lemmas));
	}

	public static void testCICounts(final CoreModel model, final String word, final PrintStream ps)
	{
		var count = LexGroupings.countsByICLemma(model).get(word);
		ps.printf("%s %d%n", word, count);
	}

	public static void testCICountsFromMap(final CoreModel model, final String word, final PrintStream ps)
	{
		var count = LexGroupings.multipleCountsByICLemma(model).get(word);
		ps.printf("%s %d%n", word, count);
	}

	public static void testCIHypermap3(final CoreModel model, String word, final PrintStream ps)
	{
		String s1 = testCIHypermapString(model, word);
		ps.printf("ci '%s'%n", word);
		ps.println(s1);

		word = word.toLowerCase(Locale.ENGLISH);
		String s2 = testCIHypermapString(model, word);
		ps.printf("ci '%s'%n", word);
		ps.println(s2);

		word = word.toUpperCase(Locale.ENGLISH);
		String s3 = testCIHypermapString(model, word);
		ps.printf("ci '%s'%n", word);
		ps.println(s3);

		assertEquals(s1, s2);
		assertEquals(s2, s3);
	}

	public static void testCILexesFor3(final CoreModel model, String word, final PrintStream ps)
	{
		String s1 = testCILexesString(model, word);
		ps.printf("ci '%s'%n", word);
		ps.println(s1);

		word = word.toLowerCase(Locale.ENGLISH);
		String s2 = testCILexesString(model, word);
		ps.printf("ci '%s'%n", word);
		ps.println(s2);

		word = word.toUpperCase(Locale.ENGLISH);
		String s3 = testCILexesString(model, word);
		ps.printf("ci '%s'%n", word);
		ps.println(s3);

		assertEquals(s1, s2);
		assertEquals(s2, s3);
	}

	public static void testCIHypermap(final CoreModel model, final String word, final PrintStream ps)
	{
		ps.printf("ci '%s'%n", word);
		ps.println(testCIHypermapString(model, word));
	}

	public static void testCILexes(final CoreModel model, final String word, final PrintStream ps)
	{
		ps.printf("ci '%s'%n", word);
		ps.println(testCILexesString(model, word));
	}

	public static void testCILexesFor(final CoreModel model, final String word, final PrintStream ps)
	{
		ps.printf("ci '%s'%n", word);
		ps.println(testCILexesForWordString(model, word));
	}

	public static String testCIHypermapString(final CoreModel model, final String word)
	{
		StringWriter sw = new StringWriter();
		final var hyperMap = LexGroupings.byLCLemma(model);
		assert hyperMap != null;
		final var map = hyperMap.get(word.toLowerCase(Locale.ENGLISH));
		map.keySet().forEach(cs -> {
			sw.write(String.format("\tcs '%s'%n", cs));
			map.get(cs) //
					.forEach(s -> sw.write(String.format("\t\t%s%n", s)));
		});
		return sw.toString();
	}

	public static String testCILexesString(final CoreModel model, final String word)
	{
		StringWriter sw = new StringWriter();
		final var map = LexGroupings.lexesByLCLemma(model);
		assert map != null;
		final var lexes = map.get(word.toLowerCase(Locale.ENGLISH));
		assert lexes != null;
		return lexesToString(lexes);
	}

	public static String testCILexesForWordString(final CoreModel model, final String word)
	{
		StringWriter sw = new StringWriter();
		final var lexes = LexGroupings.lexesForLCLemma(model, word);
		assert lexes != null;
		return lexesToString(lexes);
	}

	public static void testCISensesGroupingByLCLemmaAndPos(final CoreModel model, final String word, final char pos, final PrintStream ps)
	{
		ps.printf("ci '%s' %s%n", word, pos);
		ps.println(testCISensesGroupingByLCLemmaAndPosString(model, word, pos));
	}

	public static void testCISensesGroupingByLCLemma(final CoreModel model, final String word, final PrintStream ps)
	{
		ps.printf("ci '%s'%n", word);
		ps.println(testCISensesGroupingByLCLemmaString(model, word));
	}

	private static String testCISensesGroupingByLCLemmaString(final CoreModel model, final String word)
	{
		final var senses = sensesForLCLemma(model, word);
		return sensesToString(senses);
	}

	private static String testCISensesGroupingByLCLemmaAndPosString(final CoreModel model, final String word, final char pos)
	{
		final var senses = sensesForLCLemmaAndPos(model, word, pos);
		return sensesToString(senses);
	}

	private static String sensesToString(final List<Sense> senses)
	{
		if (senses == null || senses.isEmpty())
		{
			return "\t<none>";
		}
		StringWriter sw = new StringWriter();
		senses.forEach(sense -> {
			sw.write(String.format("\t%s%n", sense));
		});
		return sw.toString();
	}

	private static String lexesToString(final List<Lex> lexes)
	{
		if (lexes == null || lexes.isEmpty())
		{
			return "\t<none>";
		}
		StringWriter sw = new StringWriter();
		lexes.forEach(lex -> {
			sw.write(String.format("\t%s%n", lex));
		});
		return sw.toString();
	}
}

File: src/test/java/org/oewntk/model/LibTestModelLexGroups.kt
/*
 * Copyright (c) 2021-2024. Bernard Bou.
 */
package org.oewntk.model

import org.junit.Assert
import org.oewntk.model.LibLexGroupings.lemmasByLCLemmaHavingMultipleCount
import org.oewntk.model.LibLexGroupings.countsByLCLemma
import org.oewntk.model.LibLexGroupings.findLemmasFor
import org.oewntk.model.LibLexGroupings.multipleCountsByICLemma
import org.oewntk.model.SenseGroupings.sensesForLCLemma
import org.oewntk.model.SenseGroupings.sensesForLCLemmaAndPos
import org.oewntk.model.TestUtils.lexHypermapForLemmaToString
import org.oewntk.model.TestUtils.lexesToString
import org.oewntk.model.TestUtils.sensesToString
import java.io.PrintStream
import java.util.Locale

object LibTestModelLexGroups {

    fun testCIMultipleAll(model: CoreModel, ps: PrintStream) {
        model.lexes.lemmasByLCLemmaHavingMultipleCount()
            .forEach { (u: String?, cs: Set<String>) ->
                ps.println("$u ${cs.joinToString(separator = ",", prefix = "{", postfix = "}")}")
            }
    }

    fun testCILemmas(model: CoreModel, lemma: Lemma, ps: PrintStream) {
        val lemmas = model.lexes.findLemmasFor(lemma)
        ps.println("$lemma ${lemmas.joinToString(separator = ",", prefix = "{", postfix = "}")}")
    }

    fun testCICounts(model: CoreModel, lemma: Lemma, ps: PrintStream) {
        val count = model.lexes.countsByLCLemma()[lemma]
        ps.printf("%s %d%n", lemma, count)
    }

    fun testCICountsFromMap(model: CoreModel, lemma: Lemma, ps: PrintStream) {
        val count = model.lexes.multipleCountsByICLemma()[lemma]
        ps.printf("%s %d%n", lemma, count)
    }

    fun testCILexesFor3(model: CoreModel, lemma: Lemma, ps: PrintStream) {
        var lemma2 = lemma
        val s1 = testCILexesFor(model, lemma2)
        ps.printf("ci '%s'%n", lemma2)
        ps.println(s1)

        lemma2 = lemma2.lowercase(Locale.ENGLISH)
        val s2 = testCILexesFor(model, lemma2)
        ps.printf("ci '%s'%n", lemma2)
        ps.println(s2)

        lemma2 = lemma2.uppercase()
        val s3 = testCILexesFor(model, lemma2)
        ps.printf("ci '%s'%n", lemma2)
        ps.println(s3)

        Assert.assertEquals(s1, s2)
        Assert.assertEquals(s2, s3)
    }

    fun testCILexesFor(model: CoreModel, lemma: Lemma, ps: PrintStream) {
        ps.printf("ci '%s'%n", lemma)
        ps.println(testCILexesFor(model, lemma))
    }

    private fun testCILexesFor(model: CoreModel, lemma: Lemma): String {
        val lexes = checkNotNull(model.lexIgnoreCaseFinder(lemma))
        return lexesToString(lexes)
    }

    fun testCISensesGroupingByLCLemmaAndPos(model: CoreModel, lemma: Lemma, category: Category, ps: PrintStream) {
        ps.printf("ci '%s' %s%n", lemma, category)
        ps.println(testCISensesGroupingByLCLemmaAndPosString(model, lemma, category))
    }

    fun testCISensesGroupingByLCLemma(model: CoreModel, lemma: Lemma, ps: PrintStream) {
        ps.printf("ci '%s'%n", lemma)
        ps.println(testCISensesGroupingByLCLemmaString(model, lemma))
    }

    private fun testCISensesGroupingByLCLemmaString(model: CoreModel, lemma: Lemma): String {
        val senses = sensesForLCLemma(model.senses, lemma)
        return sensesToString(senses)
    }

    private fun testCISensesGroupingByLCLemmaAndPosString(model: CoreModel, lemma: Lemma, category: Category): String {
        val senses = sensesForLCLemmaAndPos(model.senses, lemma, category)
        return sensesToString(senses)
    }

    fun testCIHypermap(lexHyperMap: Map<String, Map<String, Collection<Lex>>>, lemma: Lemma, ps: PrintStream) {
        ps.printf("ci '%s'%n", lemma)
        ps.println(testCIHypermapString(lexHyperMap, lemma))
    }

    fun testCIHypermap3(lexHyperMap: Map<String, Map<String, Collection<Lex>>>, lemma: Lemma, ps: PrintStream) {
        var lemma2 = lemma
        val s1 = testCIHypermapString(lexHyperMap, lemma2)
        ps.printf("ci '%s'%n", lemma2)
        ps.println(s1)

        lemma2 = lemma2.lowercase(Locale.ENGLISH)
        val s2 = testCIHypermapString(lexHyperMap, lemma2)
        ps.printf("ci '%s'%n", lemma2)
        ps.println(s2)

        lemma2 = lemma2.uppercase()
        val s3 = testCIHypermapString(lexHyperMap, lemma2)
        ps.printf("ci '%s'%n", lemma2)
        ps.println(s3)

        Assert.assertEquals(s1, s2)
        Assert.assertEquals(s2, s3)
    }

    private fun testCIHypermapString(lexHyperMap: Map<Lemma, Map<Lemma, Collection<Lex>>>, lemma: Lemma): String {
        return lexHypermapForLemmaToString(lexHyperMap, lemma)
    }
}
